src/com/maddyhome/idea/vim/group/ChangeGroup.java

```diff
@@ -452,7 +452,9 @@ private void initInsert(@NotNull Editor editor, @NotNull DataContext context, @N
       if (mode == CommandState.Mode.REPLACE) {
         processInsert(editor, context);
       }
-      state.pushState(mode, CommandState.SubMode.NONE, MappingMode.INSERT);
+      if (state.getSubMode() != CommandState.SubMode.SINGLE_COMMAND || mode != CommandState.Mode.INSERT) {
```

Comment:
Yes, this is a good question. I'll try to explain:
First, this is a simplified version of `!(state.getSubMode() == CommandState.SubMode.SINGLE_COMMAND && mode == CommandState.Mode.INSERT)` which means __NOT (now is single_command && new mode is insert_mode)__. The difference is when we try to enter not insert mode, for example, replace mode.

Case:
 1) Enter insert mode
 2) Enter SINGLE_COMMAND
 3) Press `R` and replace some symbols
 4) Press `esc`
With my condition after pressing `esc`, insert mode will be return.
With your condition after pressing `esc`, normal mode will be entered.

From my point of view, it is more logical that the insert mode is entered after `esc`. BUT surprisingly original VIM goes into normal mode, so your suggestion is more suitable.
But this is also not so easy, because if you do it like this, caret will be mixed. Insert mode will have block caret and normal mode will have normal caret. There is no such bug in original vim.

Anyway, if you think, that this is the required behaviour, I agree with this change.